libdnf/rpm/nevra.cpp

```diff
@@ -19,51 +19,216 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 
 #include "libdnf/rpm/nevra.hpp"
 
-#include <regex>
 
 namespace libdnf::rpm {
 
 
-#define PKG_NAME    "([^:(/=<> ]+)"
-#define PKG_EPOCH   "(([\\][*?0-9]+):)?"
-#define PKG_VERSION "([^-:(/=<> ]+)"
-#define PKG_RELEASE PKG_VERSION
-#define PKG_ARCH    "([^-:.(/=<> ]+)"
-
-// clang-format off
-static const std::regex NEVRA_FORM_REGEX[]{
-    std::regex("^" PKG_NAME "-" PKG_EPOCH PKG_VERSION "-" PKG_RELEASE "\\." PKG_ARCH "$"),
-    std::regex("^" PKG_NAME "-" PKG_EPOCH PKG_VERSION "-" PKG_RELEASE          "()"  "$"),
-    std::regex("^" PKG_NAME "-" PKG_EPOCH PKG_VERSION        "()"              "()"  "$"),
-    std::regex("^" PKG_NAME      "()()"      "()"            "()"     "\\." PKG_ARCH "$"),
-    std::regex("^" PKG_NAME      "()()"      "()"            "()"              "()"  "$")
-};
-// clang-format on
-
 static const std::vector<Nevra::Form> PKG_SPEC_FORMS{
     Nevra::Form::NEVRA, Nevra::Form::NA, Nevra::Form::NAME, Nevra::Form::NEVR, Nevra::Form::NEV};
 
 const std::vector<Nevra::Form> & Nevra::get_default_pkg_spec_forms() {
     return PKG_SPEC_FORMS;
 }
 
-bool Nevra::parse(const std::string & nevra_str, Form form) {
-    enum { NAME = 1, EPOCH = 3, VERSION = 4, RELEASE = 5, ARCH = 6, _LAST_ };
-
-    std::smatch match;
-    if (!std::regex_match(
-            nevra_str, match, NEVRA_FORM_REGEX[static_cast<std::underlying_type<Form>::type>(form) - 1])) {
-        return false;
+std::vector<Nevra> Nevra::parse(const std::string & nevra_str, const std::vector<Form> & forms) {
+    std::vector<Nevra> result;
+    const char * evr_delim = nullptr;
+    const char * epoch_delim = nullptr;
+    const char * release_delim = nullptr;
+    const char * first_dot = nullptr;
+    const char * arch_delim = nullptr;
+    const char * end;
+
+    // parse nevra
+    const char * nevra_pattern = nevra_str.c_str();
+    for (end = nevra_pattern; *end != '\0'; ++end) {
+        if (*end == '-') {
+            evr_delim = release_delim;
+            release_delim = end;
+        } else if (*end == '.') {
+            arch_delim = end;
+            if (first_dot == nullptr) {
+                first_dot = end;
```

Comment:
May be you have forgotten something, because `first_dot` is assigned but never used.